WeakAuras/WeakAuras.lua

```diff
@@ -3447,52 +3447,89 @@ function WeakAuras.ValidateUniqueDataIds(silent)
 end
 
 function WeakAuras.SyncParentChildRelationships(silent)
-  local childToParent = {};
-  local parentToChild = {};
+  -- 1. Find all auras where data.parent ~= nil or data.controlledChildren ~= nil
+  --    If an aura has both, then remove data.parent
+  --    If an aura has data.parent which doesn't exist, then remove data.parent
+  --    If an aura has data.parent which doesn't have data.controledChildren, then remove data.parent
+  -- 2. For each aura with data.controlledChildren, iterate through the list of children and remove entries where:
+  --    The child doesn't exist in the database
+  --    The child ID is duplicated in data.controlledChildren (only the first will be kept)
+  --    The child's data.parent points to a different parent
+  --    Otherwise, mark the child as having a valid parent relationship
+  -- 3. For each aura with data.parent, remove data.parent if it was not marked to have a valid relationship in 2.
+  local parents = {}
+  local children = {}
+  local childHasParent = {}
   for id, data in pairs(db.displays) do
-    if(data.parent) then
-      if(data.controlledChildren) then
-        if not(silent) then
-          print("|cFF8800FFWeakAuras|r detected desynchronization in saved variables:", id, "has both child and parent");
+    if data.parent then
+      if data.controlledChildren then
+        if not silent then
+          prettyPrint("detected corruption in saved variables: "..id.." has both child and parent.")
         end
         -- A display cannot have both children and a parent
-        data.parent = nil;
-      elseif(db.displays[data.parent] and db.displays[data.parent].controlledChildren) then
-        childToParent[id] = data.parent;
-        parentToChild[data.parent] = parentToChild[data.parent] or {};
-        parentToChild[data.parent][id] = true;
-      else
+        data.parent = nil
```

Comment:
In this code path you probably want to add the aura to parents, so that the later consistency checks on data.controlledChildren are done.

Reply:
Yup, that sounds right.